scheduler/scheduler.go

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 2019-present Future Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package scheduler
+
+import (
+	"log"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/future-architect/gcp-instance-scheduler/model"
+	"github.com/future-architect/gcp-instance-scheduler/notice"
+	"github.com/future-architect/gcp-instance-scheduler/operator"
+	"github.com/future-architect/gcp-instance-scheduler/report"
+
+	"github.com/hashicorp/go-multierror"
+	"golang.org/x/net/context"
+)
+
+// Operation target label name
+const TargetLabel = "state-scheduler"
+
+// API call interval
+const ShutdownInterval = 50 * time.Millisecond
+
+type SubscribedMessage struct {
+	Command string `json:"command"`
+}
+
+type ShutdownOptions struct {
+	PubMessage    SubscribedMessage
+	Project       string
+	Timeout       time.Duration
+	SlackAPIToken string
+	SlackChannel  string
+	SlackEnable   bool
+}
+
+func NewSchedulerOptions(msg SubscribedMessage, projectID, timeout, slackToken, slackChannel, slackEnable string) (*ShutdownOptions, error) {
+	t := time.Duration(60) * time.Second
+	if len(timeout) != 0 {
+		tm, err := strconv.Atoi(timeout)
+		if err != nil {
+			return nil, err
+		}
+		t = time.Duration(tm)
+	}
+	slackFlag := false
+	if slackEnable == "true" {
+		slackFlag = true
+	}
+	if slackToken == "" {
+		slackToken = os.Getenv("SLACK_API_TOKEN")
+	}
+	if slackChannel == "" {
+		slackChannel = os.Getenv("SLACK_CHANNEL")
+	}
```

Comment:
🤔I do not want to touch environment variables in the constructor.
Please get env parameter at side of caller.